### PR TITLE
Feature/context

### PR DESCRIPTION
**Pull request by aradyaron**

### CODE REVIEW

File: include/octo-logger-cpp/context-info.hpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
+/**
+ * @file context-info.hpp
```
Comment (SheyDenis) [inline include/octo-logger-cpp/context-info.hpp:2]: Please move the definitions to a src file.

File: include/octo-logger-cpp/log.hpp [hunk 2]
```diff
@@ -71,21 +78,19 @@ class Log
...
         if (stream_)
         {
             *stream_ << fmt::format(fmt, std::forward<T>(args)...);
         }
+        return *this;
     }
 
-    template <typename... Args>
-    void formattedf(char const* fmt, Args... args)
+    Log& with_context(ContextInfo context_info)
```
Comment (SheyDenis) [inline include/octo-logger-cpp/log.hpp:90]: Please replace this with `Log error(std::string_view extra_identifier, ContextInfo context_info) const;` (and other log levels).

File: include/octo-logger-cpp/log.hpp [hunk 3]
```diff
@@ -71,21 +78,19 @@ class Log
...
     {
         if (stream_)
         {
             *stream_ << fmt::format(fmt, std::forward<T>(args)...);
         }
+        return *this;
     }
 
-    template <typename... Args>
-    void formattedf(char const* fmt, Args... args)
```
Comment (SheyDenis) [inline include/octo-logger-cpp/log.hpp:83]: Why was this removed?

File: src/manager.cpp [hunk 4]
```diff
@@ -127,12 +127,14 @@ void Manager::stop(bool discard)
     }
 }
 
-void Manager::dump(const Log& log, const std::string& channel_name, Logger::ContextInfo const& context_info)
+void Manager::dump(const Log& log, const std::string& channel_name, ContextInfo const& context_info)
 {
+    // Copy shared pointer in order to allow update without locking on set_global_context_info
+    auto context_info_handle(global_context_info_);
```
Comment (SheyDenis) [inline src/manager.cpp:133]: This PR adds rather a lot of copies.

File: unittests/include/dummy-sink.hpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        .context_info = context_info,
+        .channel_name = channel.channel_name(),
+        .log_level = Log::level_to_string(log.log_level())
+      });
+    }
+};
+} // namespace octo::logger::unittests
+
+#endif
+#endif
```
Comment (SheyDenis) [inline unittests/include/dummy-sink.hpp:59]: Missing newline at EOF.

File: unittests/src/logging-tests.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        REQUIRE(dummy_sink_->last_log().context_info.contains("key5"));
+    }
+
+    SECTION("With New Context Info")
+    {
+        ContextInfo ci{{"key5", "value5"}};
+        manager.set_global_context_info(std::move(ci));
+        Logger logger("logging-tests");
+        logger.add_context_keys({{"key3", "value3"}});
+        logger.info("Test log").with_context({{"key4", "value4"}});
```
Comment (SheyDenis) [inline unittests/src/logging-tests.cpp:130]: Missing test that using the same logger again would not contain the ContextInfo passed in `with_context()`.

File: unittests/src/logging-tests.cpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        REQUIRE_FALSE(dummy_sink_->last_log().context_info.contains("key1"));
+        REQUIRE_FALSE(dummy_sink_->last_log().context_info.contains("key2"));
+        REQUIRE(dummy_sink_->last_log().context_info.contains("key3"));
+        REQUIRE(dummy_sink_->last_log().context_info.contains("key4"));
+        REQUIRE(dummy_sink_->last_log().context_info.contains("key5"));
+    }
+}
```
Comment (SheyDenis) [inline unittests/src/logging-tests.cpp:137]: Missing newline at EOF.

File: unittests/src/logging-tests.cpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        ContextInfo ci{{"key1", "value1"}, {"key2", "value2"}};
+        manager.set_global_context_info(std::move(ci));
+        Logger logger("logging-tests");
+        logger.add_context_keys({{"key3", "value3"}, {"key4", "value4"}});
+        logger.info("Test log").with_context({{"key5", "value5"}});
+        REQUIRE(dummy_sink_->last_log().context_info.contains("key1"));
+        REQUIRE(dummy_sink_->last_log().context_info.contains("key2"));
+        REQUIRE(dummy_sink_->last_log().context_info.contains("key3"));
+        REQUIRE(dummy_sink_->last_log().context_info.contains("key4"));
+        REQUIRE(dummy_sink_->last_log().context_info.contains("key5"));
```
Comment (SheyDenis) [inline unittests/src/logging-tests.cpp:121]:
```suggestion
        REQUIRE_THAT(dummy_sink_->last_log().context_info, ContextInfoEquals(ContextInfo({{ {"key1", "value1"}, {"key2", "value2"}, {"key3", "value3"}, {"key4", "value4"} }})));
```

File: src/logger.cpp [hunk 9]
```diff
@@ -16,7 +16,7 @@ namespace octo::logger
 {
 void Logger::dump_log(const Log& log) const
 {
-    Manager::instance().dump(log, channel_view_.channel().channel_name(), context_info_);
+    Manager::instance().dump(log, channel_view_.channel().channel_name(), log.context_info() + context_info_);
```
Comment (SheyDenis) [inline src/logger.cpp:19]: A lot of allocations, can this be refactored?

File: include/octo-logger-cpp/context-info.hpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+#include <sstream>
+
+
+namespace octo::logger
+{
+
+class ContextInfo
+{
+public:
+    typedef std::unordered_map<std::string_view, std::string> ContextInfoType;
```
[the remark this answers is not in the document]
Comment (aradyaron) [reply include/octo-logger-cpp/context-info.hpp]: added

File: include/octo-logger-cpp/context-info.hpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    [[nodiscard]] bool empty() const
+    {
+        return context_info_.empty();
+    }
+
+    [[nodiscard]] bool contains(const std::string_view& key) const
+    {
+        return context_info_.find(key) != context_info_.end();
```
[the remark this answers is not in the document]
Comment (aradyaron) [reply include/octo-logger-cpp/context-info.hpp]: fixed

File: include/octo-logger-cpp/context-info.hpp [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    void clear()
+    {
+        return context_info_.clear();
+    }
+
+    ContextInfoType::iterator begin()
```
[the remark this answers is not in the document]
Comment (aradyaron) [reply include/octo-logger-cpp/context-info.hpp]: added

File: include/octo-logger-cpp/context-info.hpp (hunk 1 above)
Comment (aradyaron) [reply include/octo-logger-cpp/context-info.hpp:2]: done

File: include/octo-logger-cpp/log.hpp (hunk 2 above)
Comment (aradyaron) [reply include/octo-logger-cpp/log.hpp]: reverted

File: include/octo-logger-cpp/log.hpp (hunk 3 above)
Comment (aradyaron) [reply include/octo-logger-cpp/log.hpp:83]: reverted

File: src/manager.cpp (hunk 4 above)
Comment (aradyaron) [reply src/manager.cpp:133]: indeed :)

File: unittests/src/logging-tests.cpp (hunk 6 above)
Comment (aradyaron) [reply unittests/src/logging-tests.cpp]: added

File: src/logger.cpp [hunk 13]
```diff
@@ -34,74 +34,72 @@ Channel& Logger::editable_logger_channel()
...
     throw std::runtime_error("No log level");
 }
 
-Logger::ContextInfo const& Logger::context_info() const
+ContextInfo const& Logger::context_info() const
 {
     return context_info_;
 }
 
 void Logger::add_context_key(std::string_view key, std::string value)
```
Comment (SheyDenis) [inline src/logger.cpp:94]:
```suggestion
void Logger::add_context_key(ContextInfo::ContextInfoKey key, ContextInfo::ContextInfoValue value)
```

File: include/octo-logger-cpp/context-info.hpp [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    ContextInfo(ContextInfo const&) = default;
+    ContextInfo& operator=(ContextInfo const&) = default;
+    [[nodiscard]] ContextInfo operator+(ContextInfo const& other) const;
+    ContextInfo& operator+=(ContextInfo const& other);
+    [[nodiscard]] bool operator==(ContextInfo const& other) const;
+    void update(ContextInfoKey key, ContextInfoValue value);
+    void update(ContextInfo const& other);
+    void erase(ContextInfoKey key);
+    [[nodiscard]] bool empty() const;
+    [[nodiscard]] bool contains(ContextInfoKey key) const;
```
Comment (SheyDenis) [inline include/octo-logger-cpp/context-info.hpp:53]:
```suggestion
    [[nodiscard]] bool contains(ContextInfoKey const& key) const;
```

File: include/octo-logger-cpp/context-info.hpp [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    ContextInfo(ContextInfo&&) = default;
+    ContextInfo& operator=(ContextInfo&&) = default;
+    ContextInfo(ContextInfo const&) = default;
+    ContextInfo& operator=(ContextInfo const&) = default;
+    [[nodiscard]] ContextInfo operator+(ContextInfo const& other) const;
+    ContextInfo& operator+=(ContextInfo const& other);
+    [[nodiscard]] bool operator==(ContextInfo const& other) const;
+    void update(ContextInfoKey key, ContextInfoValue value);
+    void update(ContextInfo const& other);
+    void erase(ContextInfoKey key);
```
Comment (SheyDenis) [inline include/octo-logger-cpp/context-info.hpp:51]:
```suggestion
    void erase(ContextInfoKey const& key);
```

File: include/octo-logger-cpp/log.hpp [hunk 16]
```diff
@@ -71,14 +78,15 @@ class Log
...
 
     template <typename... T>
-    void formatted(fmt::format_string<T...> fmt, T&&... args)
+    Log& formatted(fmt::format_string<T...> fmt, T&&... args)
     {
         if (stream_)
         {
             *stream_ << fmt::format(fmt, std::forward<T>(args)...);
         }
+        return *this;
```
Comment (SheyDenis) [inline include/octo-logger-cpp/log.hpp:87]: Why?

File: src/context-info.cpp [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+ *
+ */
+
+
+#include "octo-logger-cpp/context-info.hpp"
+
+namespace octo::logger
+{
+
+ContextInfo::ContextInfo(ContextInfoInitializerList init) : context_info_(init)
```
Comment (SheyDenis) [inline src/context-info.cpp:18]:
```suggestion
ContextInfo::ContextInfo(ContextInfoInitializerList init) : context_info_(std::move(init))
```

File: include/octo-logger-cpp/context-info.hpp [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+#include <sstream>
+
+
+namespace octo::logger
+{
+
+class ContextInfo
+{
+public:
+    typedef std::string_view ContextInfoKey;
```
Comment (ofiriluz) [inline include/octo-logger-cpp/context-info.hpp:27]: why not be explicit on types as those are std types?

File: include/octo-logger-cpp/context-info.hpp [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+    typedef std::string ContextInfoValue;
+    typedef std::unordered_map<ContextInfoKey, ContextInfoValue> ContextInfoType;
+    typedef std::initializer_list<ContextInfoType::value_type> ContextInfoInitializerList;
+
+private:
+    ContextInfoType context_info_;
+
+public:
+    ContextInfo() = default;
```
Comment (ofiriluz) [inline include/octo-logger-cpp/context-info.hpp:36]: why do we need this if we have initializer constructor

File: include/octo-logger-cpp/context-info.hpp [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+
+    virtual ~ContextInfo() = default;
+
+    ContextInfo(ContextInfoInitializerList init);
```
Comment (ofiriluz) [inline include/octo-logger-cpp/context-info.hpp:40]: explicit?

File: include/octo-logger-cpp/context-info.hpp [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+
+    ContextInfo(ContextInfo&&) = default;
```
Comment (ofiriluz) [inline include/octo-logger-cpp/context-info.hpp:42]: why those need to be defined? already defined by default unless deleted

File: include/octo-logger-cpp/log.hpp (hunk 16 above)
Comment (ofiriluz) [inline include/octo-logger-cpp/log.hpp:87]: why is this change needed? the formatted is for a single ofrmatted log, that can contain all the log string in it

File: include/octo-logger-cpp/log.hpp [hunk 22]
```diff
@@ -71,14 +78,15 @@ class Log
...
     }
-
```
Comment (ofiriluz) [inline include/octo-logger-cpp/log.hpp:81]: fix formattings?

File: include/octo-logger-cpp/manager.hpp [hunk 23]
```diff
@@ -61,11 +63,14 @@ class Manager
     void configure(const ManagerConfigPtr& config, bool clear_old_sinks = true);
     void terminate();
     void stop(bool discard = false);
-    void dump(const Log& log, const std::string& channel_name, Logger::ContextInfo const& context_info);
+    void dump(const Log& log, const std::string& channel_name, ContextInfo const& context_info);
     void clear_sinks();
     void clear_channels();
     void restart_sinks() noexcept;
     const Logger& global_logger() const;
+    // @brief get the global context info
```
Comment (ofiriluz) [inline include/octo-logger-cpp/manager.hpp:71]: remove brief, not really any useful info

File: src/logger.cpp (hunk 9 above)
Comment (ofiriluz) [inline src/logger.cpp:19]: This will cause a dropdown in performance
why not put the context info on a channel instead of each log?
or why not pass both the local and global context info's as references, and just use them on hte dump itself

File: src/manager.cpp (hunk 4 above)
Comment (ofiriluz) [inline src/manager.cpp:133]: why is this needed again?

**Review by ofiriluz — COMMENTED** [review]
This is a performance degredatation, please find better ways of to move ./ ref things.

File: src/manager.cpp (hunk 4 above)
Comment (aradyaron) [reply src/manager.cpp:133]: thread safe access in case concurrently used and set.
this will cause the pointer to not be deleted, and the map to remain valid until the function exit

File: include/octo-logger-cpp/context-info.hpp (hunk 20 above)
Comment (aradyaron) [reply include/octo-logger-cpp/context-info.hpp:40]: Implicit, nicer usage

File: include/octo-logger-cpp/context-info.hpp (hunk 18 above)
Comment (aradyaron) [reply include/octo-logger-cpp/context-info.hpp:27]: @SheyDenis request, I don't care wither way

File: include/octo-logger-cpp/context-info.hpp (hunk 21 above)
Comment (aradyaron) [reply include/octo-logger-cpp/context-info.hpp:42]: I prefer to e explicit

File: include/octo-logger-cpp/context-info.hpp (hunk 19 above)
Comment (aradyaron) [reply include/octo-logger-cpp/context-info.hpp:36]: removed

File: include/octo-logger-cpp/manager.hpp (hunk 23 above)
Comment (aradyaron) [reply include/octo-logger-cpp/manager.hpp]: done

File: src/logger.cpp (hunk 9 above)
Comment (aradyaron) [reply src/logger.cpp]: not more copies, verified in testss

File: include/octo-logger-cpp/log.hpp (hunk 16 above)
Comment (aradyaron) [reply include/octo-logger-cpp/log.hpp]: reverted

File: src/manager.cpp [hunk 24]
```diff
@@ -127,12 +131,14 @@ void Manager::stop(bool discard)
...
```
Comment (ofiriluz) [inline src/manager.cpp:137]: can we maybe just use weak_ptr?

Comment (aradyaron) [reply src/manager.cpp:137]: no, we want this as owner of the pointer.
think of the following:
thread 1: writing log
thread 1: copy shared ptr
thread 2: setting new context == updating the shared ptr count
thread 1: accessing old context.

If this was not a shared pointer on this point it could be deleted

File: src/logger.cpp [hunk 25]
```diff
@@ -34,84 +34,82 @@ Channel& Logger::editable_logger_channel()
...
+Log Logger::log(Log::LogLevel level, std::string_view extra_identifier, ContextInfo context_info) const
 {
     switch (level)
     {
         case Log::LogLevel::QUIET:
-            return Log(Log::LogLevel::QUIET, extra_identifier, *this);
+            return Log(Log::LogLevel::QUIET, extra_identifier, std::move(context_info), *this);
         case Log::LogLevel::TRACE:
-            return trace(extra_identifier);
+            return trace(extra_identifier, context_info);
```
Comment (SheyDenis) [inline src/logger.cpp:74]: Missing `std::move(context_info)` for all the cases except the first?

File: src/sink.cpp [hunk 26]
```diff
@@ -83,47 +83,51 @@ static void init_context_info_impl(nlohmann::json& dst,
             throw std::runtime_error(fmt::format("Wrong context_info destination type {}", dst.type_name()));
     }
 
-    for (auto const& itr : context_info)
+    for (auto const& context_info : {log.context_info(), context_info, global_context_info})
```
Comment (SheyDenis) [inline src/sink.cpp:86]: Wouldn't this just create a copy of the 3 `ContextInfo` instances?

File: src/sink.cpp [hunk 27]
```diff
@@ -161,14 +165,15 @@ std::string Sink::formatted_log_plaintext_short(Log const& log, Channel const& c
...
-                                         Channel const& channel,
-                                         Logger::ContextInfo const& context_info) const
+std::string Sink::formatted_context_info(Log const& log, Channel const& channel, ContextInfo const& context_info, ContextInfo const& global_context_info) const
 {
     std::string context_info_str("context_info: ");
-    for (auto const& itr : context_info)
+    for (auto const& itr : {log.context_info(), context_info, global_context_info})
     {
-        context_info_str += fmt::format(FMT_STRING("[{:s}:{:s}]"), itr.first.data(), itr.second);
+        for (auto const& itr : itr)
```
Comment (SheyDenis) [inline src/sink.cpp:173]: Shadowing variable name `itr`.

File: src/sink.cpp [hunk 28]
```diff
@@ -161,14 +165,15 @@ std::string Sink::formatted_log_plaintext_short(Log const& log, Channel const& c
...
 }
 
-std::string Sink::formatted_context_info(Log const& log,
-                                         Channel const& channel,
-                                         Logger::ContextInfo const& context_info) const
+std::string Sink::formatted_context_info(Log const& log, Channel const& channel, ContextInfo const& context_info, ContextInfo const& global_context_info) const
 {
     std::string context_info_str("context_info: ");
-    for (auto const& itr : context_info)
+    for (auto const& itr : {log.context_info(), context_info, global_context_info})
```
Comment (SheyDenis) [inline src/sink.cpp:171]: Wouldn't this just create 3 copies even though they are marked `const`?

File: include/octo-logger-cpp/manager.hpp [hunk 29]
```diff
@@ -39,12 +40,12 @@ class Manager
     ManagerConfigPtr config_;
     Log::LogLevel default_log_level_;
     std::shared_ptr<Logger> global_logger_;
+    std::shared_ptr<ContextInfo const> global_context_info_;
```
Comment (SheyDenis) [inline include/octo-logger-cpp/manager.hpp:43]: Any reason for `std::shared_ptr` over `std::unique_ptr`?
Using `std::shared_ptr` implies sharing, this class is a singleton.

Comment (aradyaron) [reply include/octo-logger-cpp/manager.hpp:43]: added comment

File: src/sink.cpp (hunk 28 above)
Comment (aradyaron) [reply src/sink.cpp:171]: why?
all are by reference?

File: src/sink.cpp (hunk 26 above)
Comment (aradyaron) [reply src/sink.cpp:86]: why?
all are by reference

File: src/context-info.cpp [hunk 30]
```diff
@@ -0,0 +1,89 @@
...
+{
+    return context_info_ == other.context_info_;
+}
+
+void ContextInfo::update(ContextInfoKey key, ContextInfoValue value)
+{
+    context_info_[key] = std::move(value);
+}
+
+void ContextInfo::update(ContextInfo const& other)
```
[the remark this answers is not in the document]
Comment (aradyaron) [reply src/context-info.cpp:31]: not that I know of

File: src/aws/aws-log-system.cpp [hunk 31]
```diff
@@ -241,7 +241,8 @@ bool AwsLogSystem::allowed_sanitizer_tag(SanitizerTags tag) const
 {
     auto all_tags = std::find(allowed_sanitizer_tags_.begin(), allowed_sanitizer_tags_.end(), SanitizerTags::ALL_TAG);
     return all_tags != std::end(allowed_sanitizer_tags_) ||
-           std::find(allowed_sanitizer_tags_.begin(), allowed_sanitizer_tags_.end(), tag) != std::end(allowed_sanitizer_tags_);
+           std::find(allowed_sanitizer_tags_.begin(), allowed_sanitizer_tags_.end(), tag) !=
+               std::end(allowed_sanitizer_tags_);
```
[the remark this answers is not in the document]
Comment (aradyaron) [reply src/aws/aws-log-system.cpp:245]: not related to this Pr, keeping as is

File: src/manager.cpp (hunk 4 above)
Comment (aradyaron) [reply src/manager.cpp:137]: no more copies, just the handle

File: unittests/include/dummy-sink.hpp (hunk 5 above)
Comment (aradyaron) [reply unittests/include/dummy-sink.hpp]: done

File: src/sink.cpp (hunk 26 above)
Comment (SheyDenis) [reply src/sink.cpp:85]: I guess you're right.

File: src/sink.cpp (hunk 28 above)
Comment (SheyDenis) [reply src/sink.cpp]: I guess you're right.

File: src/manager.cpp [hunk 32]
```diff
@@ -186,6 +192,22 @@ bool Manager::mute_channel(std::string const& name)
...
 }
 
+ContextInfo const& Manager::global_context_info() const
+{
+    return *global_context_info_;
+}
+
+void Manager::replace_global_context_info(ContextInfo const context_info)
+{
+    global_context_info_ = std::make_shared<ContextInfo const>(std::move(context_info));
```
Comment (SheyDenis) [inline src/manager.cpp:202]: `context_info` is `const` qualified, `std::move` does nothing.

File: src/sink.cpp [hunk 33]
```diff
@@ -83,11 +82,14 @@ static void init_context_info_impl(nlohmann::json& dst,
             throw std::runtime_error(fmt::format("Wrong context_info destination type {}", dst.type_name()));
     }
 
-    for (auto const& itr : context_info)
+    for (auto const& context_info : {log.context_info(), context_info, global_context_info})
```
Comment (SheyDenis) [inline src/sink.cpp:85]: Variable `context_info` shadows `context_info` argument.

File: src/sink.cpp [hunk 34]
```diff
@@ -163,12 +169,16 @@ std::string Sink::formatted_log_plaintext_short(Log const& log, Channel const& c
 
 std::string Sink::formatted_context_info(Log const& log,
                                          Channel const& channel,
-                                         Logger::ContextInfo const& context_info) const
+                                         ContextInfo const& context_info,
+                                         ContextInfo const& global_context_info) const
 {
     std::string context_info_str("context_info: ");
-    for (auto const& itr : context_info)
+    for (auto const& context_info : {log.context_info(), context_info, global_context_info})
```
Comment (SheyDenis) [inline src/sink.cpp:176]: Variable name `context_info` shadows argument `context_info`.

**Review by SheyDenis — APPROVED** [review]
I guess... LGTM™

**Review by ofiriluz — APPROVED** [review]

**Review by tuvalkohl — COMMENTED** [review]
Some comments for myself

**Review by tuvalkohl — COMMENTED** [review]
A few more comments

File: include/octo-logger-cpp/aws/cloudwatch-sink.hpp [hunk 35]
```diff
@@ -130,7 +128,7 @@ class CloudWatchSink : public Sink
                    LogGroupTags log_group_tags = {});
     ~CloudWatchSink() override;
 
-    void dump(Log const& log, Channel const& channel, Logger::ContextInfo const& context_info) override;
+    void dump(Log const& log, Channel const& channel, ContextInfo const& context_info) override;
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply include/octo-logger-cpp/aws/cloudwatch-sink.hpp:131]: Done

File: include/octo-logger-cpp/context-info.hpp [hunk 36]
```diff
@@ -0,0 +1,60 @@
...
+    typedef std::string_view ContextInfoKey;
+    typedef std::string ContextInfoValue;
+    typedef std::unordered_map<ContextInfoKey, ContextInfoValue> ContextInfoType;
+    typedef std::initializer_list<ContextInfoType::value_type> ContextInfoInitializerList;
+
+  private:
+    ContextInfoType context_info_;
+
+  public:
+    virtual ~ContextInfo() = default;
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply include/octo-logger-cpp/context-info.hpp:35]: Removed virtual

File: src/manager.cpp [hunk 37]
```diff
@@ -186,6 +192,22 @@ bool Manager::mute_channel(std::string const& name)
...
 }
 
+ContextInfo const& Manager::global_context_info() const
+{
+    return *global_context_info_;
+}
+
+void Manager::replace_global_context_info(ContextInfo context_info)
+{
+    global_context_info_ = std::make_shared<ContextInfo const>(std::move(context_info));
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply src/manager.cpp:202]: Switched to atomic_store / atomic_load

File: src/manager.cpp [hunk 38]
```diff
@@ -186,6 +192,22 @@ bool Manager::mute_channel(std::string const& name)
     return true;
 }
 
+ContextInfo const& Manager::global_context_info() const
+{
+    return *global_context_info_;
+}
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply src/manager.cpp:198]: Switched to returning a `shared_ptr` (and also atomically loading), so it's now safe

File: include/octo-logger-cpp/manager.hpp [hunk 39]
```diff
@@ -61,11 +63,14 @@ class Manager
     void configure(const ManagerConfigPtr& config, bool clear_old_sinks = true);
     void terminate();
     void stop(bool discard = false);
-    void dump(const Log& log, const std::string& channel_name, Logger::ContextInfo const& context_info);
+    void dump(const Log& log, const std::string& channel_name, ContextInfo const& context_info);
     void clear_sinks();
     void clear_channels();
     void restart_sinks() noexcept;
     const Logger& global_logger() const;
+    ContextInfo const& global_context_info() const;
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply include/octo-logger-cpp/manager.hpp:71]: https://github.com/ofiriluz/octo-logger-cpp/pull/22#discussion_r1771402177

File: src/manager.cpp [hunk 40]
```diff
@@ -16,7 +16,10 @@ namespace octo::logger
 std::shared_ptr<Manager> Manager::manager_;
 std::mutex Manager::manager_init_mutex_;
 
-Manager::Manager() : config_(std::make_shared<ManagerConfig>()), default_log_level_(Log::LogLevel::INFO)
+Manager::Manager()
+    : config_(std::make_shared<ManagerConfig>()),
+      default_log_level_(Log::LogLevel::INFO),
+      global_context_info_(std::make_unique<ContextInfo>())
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply src/manager.cpp:22]: Fixed

File: src/manager.cpp [hunk 41]
```diff
@@ -186,6 +192,22 @@ bool Manager::mute_channel(std::string const& name)
...
+void Manager::replace_global_context_info(ContextInfo context_info)
+{
+    global_context_info_ = std::make_shared<ContextInfo const>(std::move(context_info));
+}
+
+void Manager::update_global_context_info(ContextInfo context_info)
+{
+    context_info.update(*global_context_info_);
+    replace_global_context_info(std::move(context_info));
+}
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply src/manager.cpp:209]: Fixed

File: src/manager.cpp [hunk 42]
```diff
@@ -127,12 +131,14 @@ void Manager::stop(bool discard)
     }
 }
 
-void Manager::dump(const Log& log, const std::string& channel_name, Logger::ContextInfo const& context_info)
+void Manager::dump(const Log& log, const std::string& channel_name, ContextInfo const& context_info)
 {
+    // Copy shared pointer in order to allow update without locking on replace_global_context_info
+    auto context_info_handle(global_context_info_);
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply src/manager.cpp:137]: Switched to atomic load/store

File: src/sink.cpp [hunk 43]
```diff
@@ -163,12 +169,16 @@ std::string Sink::formatted_log_plaintext_short(Log const& log, Channel const& c
...
 {
     std::string context_info_str("context_info: ");
-    for (auto const& itr : context_info)
+    for (auto const& ci_itr : {log.context_info(), context_info, global_context_info})
     {
-        context_info_str += fmt::format(FMT_STRING("[{:s}:{:s}]"), itr.first.data(), itr.second);
+        for (auto const& itr : ci_itr)
+        {
+            context_info_str += fmt::format(FMT_STRING("[{:s}:{:s}]"), itr.first.data(), itr.second);
+        }
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply src/sink.cpp:181]: We are fine with this behaviour, added a comment to make it explicit + wrote it in the commit message.

File: src/sink.cpp [hunk 44]
```diff
@@ -27,19 +27,17 @@
 // We always want to compile these functions, which expose our internals to the unit tests.
 namespace octo::logger::unittests
 {
-void init_context_info(nlohmann::json& dst,
-                       Log const& log,
-                       Channel const& channel,
-                       Logger::ContextInfo const& context_info);
-nlohmann::json init_context_info(Log const& log, Channel const& channel, Logger::ContextInfo const& context_info);
+void init_context_info(nlohmann::json& dst, Log const& log, Channel const& channel, ContextInfo const& context_info);
+nlohmann::json init_context_info(Log const& log, Channel const& channel, ContextInfo const& context_info);
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply src/sink.cpp:31]: Done

File: include/octo-logger-cpp/manager.hpp [hunk 45]
```diff
@@ -39,12 +40,13 @@ class Manager
     ManagerConfigPtr config_;
     Log::LogLevel default_log_level_;
     std::shared_ptr<Logger> global_logger_;
+    // Shared Pointer in order to allow thread safe usage on the 'dump' method
+    std::shared_ptr<ContextInfo const> global_context_info_;
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply include/octo-logger-cpp/manager.hpp:44]: I guess we're fine with this implicit assumption...
